### ISSUE TITLE
ibis.memtable leaks memory with default (duckdb) backend

### ISSUE THREAD

**Issue by liudvikasakelis** [ibis-project/ibis#10041]
### What happened?

Creating ephemeral ibis.memtables uses memory that is (seemingly) never freed. Reproducible example: 
```
import ibis
import os
print(os.getpid())

row = {
    'a': 123,
    'b': 'asdf',
    'c': None
}

rows = [row] * 1000000

for i in range(1, 10000):
    ibis.memtable(rows).count().execute()
```
Above script ran >20 minutes on a 32GB Win11 machine before crashing with 
```
numpy.core._exceptions._ArrayMemoryError: Unable to allocate 7.63 MiB for an array with shape (1000000,) and data type uint64
```
On my machine, task manager reported process using ~20GB for a long time before crashing, so I instead recommend tracking with perfmon (Performance Monitor), and monitoring Private Bytes from Process V2 object. This metric monotonically increased until 50+GB before crashing, but I understand this can differ between systems. 

This contrived example mimics a workflow where data chunks are read, modified using Ibis API, and written out. This memory leakage imposes an artificial limit to such pipelines.

Any workarounds would be very appreciated 🙏 

### What version of ibis are you using?

9.0.0

### What backend(s) are you using, if any?

Only tested with default duckdb 0.10.3

### Relevant log output

```sh
Win11, Python 3.12.4, ibis 9.0.0, duckdb 0.10.3
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by liudvikasakelis** [comment]
An apparent workaround seems to be `ibis.set_backend('pandas')`, which maintains all memory usage counters flat. That said, Pandas is inferior to duckdb in many respects, so I'm eager to hear what others have to say

**Comment by cpcloud** [comment]
I think we can set a finalizer to call `self.con.unregister(name)` when the memtable operation operation is about to be collected.

**Comment by cpcloud** [comment]
I just tried it out locally and was able to reproduce the problem, and using the finalizer approach the problem goes away.

**Comment by cpcloud** [comment]
Using the pandas backend is definitely not an option, since it's going away in the next major version of Ibis :)

**Comment by liudvikasakelis** [comment]
I don't remember paying for a corporate support contract 👏 What a turnaround time!
Thank you very much @cpcloud , I'll be eagerly waiting for this fix release

**Comment by cpcloud** [comment]
I've been running the leaky program for about an hour now, and it's hovering right around 260-290M of memory used by the process.

**Comment by cpcloud** [comment]
Finished without issue in `1h44m56s` 🎉

**Comment by cpcloud** [comment]
Bit of a 🐰 🕳️ this one.

**TL; DR**: we're going to release a new version of Ibis (9.5) as soon as DuckDB 1.1 is released.

Since this was a bug on both sides of that coin, it will unfortunately require upgrading both Ibis and DuckDB to pick up the fixes.
